File: src/test/java/jp/skypencil/javadocky/repository/LocalStorageTest.java
package jp.skypencil.javadocky.repository;

import static org.junit.jupiter.api.Assertions.assertIterableEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;

import java.io.File;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;
import org.springframework.core.io.buffer.DataBufferFactory;
import org.springframework.core.io.buffer.DefaultDataBufferFactory;
import reactor.core.publisher.Flux;

class LocalStorageTest {
  @Test
  void test(@TempDir Path root) throws IOException {
    Storage storage = new LocalStorage(root);
    assertNull(storage.find("g", "a", "v", "index.html").block());
    DataBufferFactory factory = new DefaultDataBufferFactory();
    storage
        .write(
            "g",
            "a",
            "v",
            "index.html",
            Flux.just(factory.wrap("Hello world!".getBytes(StandardCharsets.UTF_8))))
        .block();
    File written = storage.find("g", "a", "v", "index.html").block();
    assertNotNull(written);
    assertIterableEquals(Files.readAllLines(written.toPath()), Arrays.asList("Hello world!"));
  }
}

File: src/test/kotlin/jp/skypencil/javadocky/repository/LocalStorageTest.kt
package jp.skypencil.javadocky.repository

import io.kotest.core.spec.style.FunSpec
import io.kotest.engine.spec.tempdir
import io.kotest.matchers.collections.shouldContainExactly
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import org.springframework.core.io.buffer.DataBufferFactory
import org.springframework.core.io.buffer.DefaultDataBufferFactory
import reactor.core.publisher.Flux
import java.nio.charset.StandardCharsets
import java.nio.file.Files

class LocalStorageTest : FunSpec({
    val root = tempdir()

    test("empty directory returns null") {
        val storage: Storage = LocalStorage(root.toPath())
        storage.find("g", "a", "v", "index.html").block() shouldBe null
    }

    test("written storage returns an entry") {
        val storage: Storage = LocalStorage(root.toPath())
        val factory: DataBufferFactory = DefaultDataBufferFactory()
        storage
            .write(
                "g",
                "a",
                "v",
                "index.html",
                Flux.just(factory.wrap("Hello world!".toByteArray(StandardCharsets.UTF_8)))
            )
            .block()

        val written = storage.find("g", "a", "v", "index.html").block()
        written shouldNotBe null
        Files.readAllLines(written!!.toPath()) shouldContainExactly listOf("Hello world!")
    }
})
